math/fastor/include/algebra/math/impl/fastor_transform3.hpp

```diff
@@ -0,0 +1,245 @@
+/** Algebra plugins, part of the ACTS project
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "algebra/qualifiers.hpp"
+
+// Fastor include(s).
+#ifdef _MSC_VER
+#pragma warning(disable : 4244 4701 4702)
+#endif  // MSVC
+#include <Fastor/Fastor.h>
+#ifdef _MSC_VER
+#pragma warning(default : 4244 4701 4702)
+#endif  // MSVC
+
+// System include(s).
+#include <cstddef>
+
+namespace algebra::fastor::math {
+
+/** Transform wrapper class to ensure standard API within different plugins
+ *
+ **/
+template <typename scalar_t, typename matrix_actor_t>
+struct transform3 {
+  /// @name Type definitions for the struct
+  /// @{
+
+  /// Array type used by the transform
+  template <std::size_t N>
+  using array_type = Fastor::Tensor<scalar_t, N>;
+  /// Scalar type used by the transform
+  using scalar_type = scalar_t;
+
+  /// 3-element "vector" type
+  using vector3 = array_type<3>;
+  /// Point in 3D space
+  using point3 = vector3;
+  /// Point in 2D space
+  using point2 = array_type<2>;
+
+  /// 4x4 matrix type
+  using matrix44 = Fastor::Tensor<scalar_type, 4UL, 4UL>;
+
+  /// Function (object) used for accessing a matrix element
+  using element_getter = algebra::fastor::math::element_getter<scalar_t>;
+
+  /// Size type
+  using size_type = typename matrix_actor_t::size_ty;
+
+  /// Matrix actor
+  using matrix_actor = matrix_actor_t;
+
+  /// 2D Matrix type
+  template <size_type ROWS, size_type COLS>
+  using matrix_type = typename matrix_actor::template matrix_type<ROWS, COLS>;
+
+  /// @}
+
+  /// @name Data objects
+  /// @{
+
+  matrix44 _data;
+  matrix44 _data_inv;
+
+  /// @}
+
+  /** Contructor with arguments: t, x, y, z
+   *
+   * @param t the translation (or origin of the new frame)
+   * @param x the x axis of the new frame
+   * @param y the y axis of the new frame
+   * @param z the z axis of the new frame, normal vector for planes
+   *
+   **/
+  ALGEBRA_HOST_DEVICE
+  transform3(const vector3 &t, const vector3 &x, const vector3 &y,
+             const vector3 &z, bool get_inverse = true) {
+
+    // The matrices need to be initialized to the identity matrix first.
+    _data.eye2();
```

Comment:
> Would it not be cheaper to set the data directly?

When `_data` and `_data_inv` are created, they are filled with garbage values. (This is because Fastor doesn't really have a method to create an identity matrix outright. What we must do instead is to initialize a Tensor and _then_ set it to the identity matrix.)

> Why exactly does this need to be done?

The reason it needs to be set to the identity matrix is because the last row of the 4x4 matrix is required to be `[0, 0, 0, 1]` in the constructor. FWIW, both the [Eigen](https://github.com/acts-project/algebra-plugins/blob/78b2176587d182a8c5d4d47dbdbee3eb53192c36/math/eigen/include/algebra/math/impl/eigen_transform3.hpp#L70-L73) and [SMatrix](https://github.com/acts-project/algebra-plugins/blob/78b2176587d182a8c5d4d47dbdbee3eb53192c36/math/smatrix/include/algebra/math/impl/smatrix_transform3.hpp#L63-L64) backends set `_data` and `_data_inv` to the identity matrix upon construction.